### ISSUE TITLE
Function Dependencies Causing Build To Fail

### ISSUE THREAD

**Issue by firstlinestudio** [angular/angularfire#3025]
### Version info

**Angular:** 12.2.29

**Firebase:** 9.1.2 / 9.1.3

**AngularFire:** 7.1.0 / 7.1.1

**Other (e.g. Ionic/Cordova, Node, browser, operating system):** Windows Terminal / PowerShell

### How to reproduce these conditions

**Failing test unit, Stackblitz demonstrating the problem**

**Steps to set up and reproduce**

In any terminal / powershell, run: ng deploy.  The process will run until it attempts to upload the function.  The function fails to deploy.

**Sample data and security rules**

**package.json created in dist/{PROJECT_NAME}/functions**
```
  "dependencies": {
    "firebase-admin": "^9.11.1",
    "firebase-functions": "^3.6.0",
    "{PROJECT_NAME}": "file:..\\..\\.."
  },
```

### Debug output

** Errors in Windows Terminal **

> Error when trying to deploy:
> Functions did not deploy properly.

** Error output from Google Cloud Builder Log **
"Step #1 - "build": Failure: (ID: XXXXXX) npm ERR! @angular-devkit/architect not accessible from {PROJECT_NAME}"

### Expected behavior

To deploy successfully.  Using 7.1.0-rc.4, the dist package.json does not try to access root dependencies and deploys successfully.

### Actual behavior

Project will not deploy with cloud functions

**Comment by firstlinestudio** [comment]
@jasonbdt I found (A / THE) solution, but you may not like it.  You'll have to do some downgrading so that all your dependencies can build properly.  Here's what I did:

- Nodejs 16 -> Nodejs 14.18.1 (latest)
- NPM 7.24 or 8.1 -> NPM 6.14.15
- Typescript 4.4.4 -> Typescript 4.3.5

I believe those are the main three.  I was able to keep angular (core, common, fire, material, etc..) at the latest versions.  Be sure to change the `"functionsNodeVersion": "14"` line to read "14" and not "12" or "16"; so that it will match the node version installed on your PC.

**Comment by firstlinestudio** [comment]
@jasonbdt I'm sure that is what fixed my deploy issue, but it didn't not fix the project all together.  I now have an issue with the SSR function not retrieving URLs.  Every URL except root will return a "cannot get URL /sample/url" message when trying to access through the function.  I can navigate the site just fine, but if I refresh a page -- other than the root homepage -- it will throw an error.  That is in development.  In production, the message shows `Error Forbidden: Your client does not have permission to get URL /ssr_sample/url1/url2 from this server.` .  I am still working on a solution.

**Comment by firstlinestudio** [comment]
The project is back to working fully.  `firebase-tools` was installed globally and locally (in the project).  I removed the local package since that is not needed.  I also had to update `@angular-devkit/architect` using `npm install @angular-devkit/architect@latest`.  That fixed the development build problems.  I had to disable cloud functions, then re-enable cloud functions to get the production build to work (using Google Cloud Web Console).  Finally, I removed `"region": "us-central1",` from the `angular.json` file, under the "deploy" settings.

The moral of the story is to manage dependencies properly.  I'm not sure if it is listed in the docs, but it would be useful to add "optimal" package versions somewhere (i.e. use nodejs 14, not nodejs 16) and/or have an optimal package.json available for the latest build.  I recommend NOT doing a general `npm update` on the root package.json, since having the latest versions of all packages doesn't always seem to work as well as expected.

**Comment by jasonbdt** [comment]
@firstlinestudio of course, if you have installed firebase-tools globally you can remove the local package. I guess the developers doesn't suppose everyone to have firebase-tools installed globally, thats why they added it as local package which makes sense for me.

As far as I know `@angular-devkit/architext` have to be in a version which mirrors the version of the core like shown below:

```json
"devDependencies": {
    "@angular-devkit/architect": "^0.1401.0",
    "@angular-devkit/build-angular": "^14.1.0",
}
```

The `region` option inside `angular.json` is only interesting if you want to use another region to upload your cloud function. If you are fine using us-central1, you can remove this as it's the default region.

When I use `ng deploy` the CLI creates a `/dist/functions` directory for me. Inside this directory there is a `package.json`. For any reason, @angular/fire is not even writing `"[PACKAGE NAME]": "file:..\..\.."` under dependencies section. My package.json generated by the CLI looks like this:

```json
{
  "name": "functions",
  "description": "Angular Universal Application",
  "main": "index.js",
  "scripts": {
    "start": "firebase functions:shell"
  },
  "engines": {
    "node": "14"
  },
  "dependencies": {
    "firebase-admin": "^9.11.1",
    "firebase-functions": "^3.6.0"
  },
  "devDependencies": {},
  "private": true
}
```

If all files created by the CLI for uploading to Cloud Functions, the command tries to install the dependencies named inside the `/dist/functions/package.json` file. After that the upload begins and our cloud function should be updated. The upload process will throw an error, because the required angular packages can not be found.

Because of this, I have to mirror the dependencies section of my root package.json into `/dist/functions/package.json`, move into this directory and reinstall the packages manually by using `npm install`. After that I can move back into my root directory with the terminal and run `firebase deploy`. This workaround works for me.

I have also copied this repository and build a custom version of it with some code changes to make `ng deploy` work again. After some trial and error I found a solution. I will do some more testing and if everything still works as expected, I will make a pull request to fix this.


Sincerly
Jason

**Comment by sidsaxena0** [comment]
> @firstlinestudio of course, if you have installed firebase-tools globally you can remove the local package. I guess the developers doesn't suppose everyone to have firebase-tools installed globally, thats why they added it as local package which makes sense for me.
> 
> As far as I know `@angular-devkit/architext` have to be in a version which mirrors the version of the core like shown below:
> 
> ```json
> "devDependencies": {
>     "@angular-devkit/architect": "^0.1401.0",
>     "@angular-devkit/build-angular": "^14.1.0",
> }
> ```
> 
> The `region` option inside `angular.json` is only interesting if you want to use another region to upload your cloud function. If you are fine using us-central1, you can remove this as it's the default region.
> 
> When I use `ng deploy` the CLI creates a `/dist/functions` directory for me. Inside this directory there is a `package.json`. For any reason, @angular/fire is not even writing `"[PACKAGE NAME]": "file:..\..\.."` under dependencies section. My package.json generated by the CLI looks like this:
> 
> ```json
> {
>   "name": "functions",
>   "description": "Angular Universal Application",
>   "main": "index.js",
>   "scripts": {
>     "start": "firebase functions:shell"
>   },
>   "engines": {
>     "node": "14"
>   },
>   "dependencies": {
>     "firebase-admin": "^9.11.1",
>     "firebase-functions": "^3.6.0"
>   },
>   "devDependencies": {},
>   "private": true
> }
> ```
> 
> If all files created by the CLI for uploading to Cloud Functions, the command tries to install the dependencies named inside the `/dist/functions/package.json` file. After that the upload begins and our cloud function should be updated. The upload process will throw an error, because the required angular packages can not be found.
> 
> Because of this, I have to mirror the dependencies section of my root package.json into `/dist/functions/package.json`, move into this directory and reinstall the packages manually by using `npm install`. After that I can move back into my root directory with the terminal and run `firebase deploy`. This workaround works for me.
> 
> I have also copied this repository and build a custom version of it with some code changes to make `ng deploy` work again. After some trial and error I found a solution. I will do some more testing and if everything still works as expected, I will make a pull request to fix this.
> 
> Sincerly Jason

Have you been able to solve it in your new repository, this issue is too old and it's sad there is still no solution for this.

**Comment by jasonbdt** [comment]
> Have you been able to solve it in your new repository, this issue is too old and it's sad there is still no solution for this.

Yes, I was able to find a solution but sadly I had no time to test it properly to check if my solution would break anything else.

**Comment by nicky-lenaers** [comment]
@jasonbdt Would you be able to share your solution by any chance?

**Comment by armando-navarro** [comment]
Thanks for the report, and apologies for the very long delay here.

This one is from the AngularFire 7 / Angular 12 era, and the part of the deploy schematic that generates the functions `package.json` has been rewritten since. In your report the generated `functions/package.json` added a `file:..\..\..` entry pointing at the whole app, so Cloud Build tried to install the root project and failed on a dev dependency (`@angular-devkit/architect not accessible`). I checked the current generator against a freshly scaffolded SSR app, and it no longer does that: it writes `firebase-admin`/`firebase-functions` and (for a server build that externalizes packages) those external deps, with no self-reference to your app. So the specific failure you hit should not recur on a current version.

One caveat worth flagging: that deploy path was written for the older server-builder model, and the current `application` builder used by new SSR apps does not create the `server` target it reads, so `ng deploy` to Cloud Functions may need separate attention on current setups. If you are still deploying functions this way, could you retest on a current version and, if it fails, share the generated `dist/<project>/functions/package.json` and the build error? Then I can pin down whether it is the old bug (it should not be) or this newer builder mismatch.
